src/context/provided_types.rs

```diff
@@ -0,0 +1,92 @@
+use std::{marker::PhantomData, mem, sync::OnceLock};
+
+use better_any::TidAble;
+
+use crate::FileRunContext;
+
+pub trait FromFileRunContextInstanceProvider<'a>: Sized {
+    type Parent: FromFileRunContextInstanceProviderFactory<Provider<'a> = Self>;
+
+    fn get<T: FromFileRunContext<'a> + TidAble<'a>>(
+        &self,
+        file_run_context: FileRunContext<'a, '_, Self::Parent>,
+    ) -> Option<&T>;
+}
+
+pub trait FromFileRunContextInstanceProviderFactory: Send + Sync {
+    type Provider<'a>: FromFileRunContextInstanceProvider<'a, Parent = Self>;
+
+    fn create<'a>(&self) -> Self::Provider<'a>;
+}
+
+pub trait FromFileRunContext<'a> {
+    fn from_file_run_context(
+        file_run_context: FileRunContext<'a, '_, impl FromFileRunContextInstanceProviderFactory>,
+    ) -> Self;
+}
+
+mod _sealed {
+    use super::*;
+
+    pub trait Sealed<'a> {}
+
+    impl<'a, T1: FromFileRunContext<'a>> Sealed<'a> for (T1,) {}
+}
+
+pub trait FromFileRunContextProvidedTypes<'a>: _sealed::Sealed<'a> {
+    type OnceLockStorage: FromFileRunContextProvidedTypesOnceLockStorage<'a>;
+
+    // fn len(&self) -> usize;
+    // fn get_index<'a, T: FromFileRunContext<'a>>(&self) -> Option<usize>;
+    fn once_lock_storage() -> Self::OnceLockStorage;
+}
+
+impl<'a, T1> FromFileRunContextProvidedTypes<'a> for (T1,)
+// where
+//     T1: FromFileRunContext<'a> + for<'b> TidAble<'b>,
+where
+    T1: FromFileRunContext<'a> + TidAble<'a>,
+{
+    type OnceLockStorage = FromFileRunContextProvidedTypesOnceLockStorageEnum<'a, T1>;
+
+    fn once_lock_storage() -> Self::OnceLockStorage {
+        FromFileRunContextProvidedTypesOnceLockStorageEnum::One(Default::default(), PhantomData)
+    }
+}
+
+pub trait FromFileRunContextProvidedTypesOnceLockStorage<'a> {
+    // fn get<T: FromFileRunContext<'a> + for<'b> TidAble<'b>>(
+    fn get<T: FromFileRunContext<'a> + TidAble<'a>>(
+        &self,
+        file_run_context: FileRunContext<'a, '_, impl FromFileRunContextInstanceProviderFactory>,
+    ) -> Option<&T>;
+}
+
+pub enum FromFileRunContextProvidedTypesOnceLockStorageEnum<'a, T1> {
+    One(OnceLock<T1>, PhantomData<&'a ()>),
```

Comment:
I think the `PhantomData` was necessary to convince it that the lifetime `'a` is actually "used"

Reply:
Ah yeah, I've never used `PhantomData` for that, but that makes sense. I wouldn't have thought to use a `()` as its type parameter—that's nice.